File: backend/app/routes/relatorios.py
from fastapi import APIRouter, Depends, Query
from typing import Optional
from datetime import date
from app.db import supabase
from app.auth import get_current_user

router = APIRouter()

@router.get("/geral")
async def relatorio_geral(
    draw_id: Optional[int] = None,
    data_inicio: Optional[str] = None,
    data_fim: Optional[str] = None,
    user=Depends(get_current_user)
):
    q = supabase.table("apostas").select("amount, prize_amount, status, bet_type, user_id, draw_id") \
        .neq("status", "cancelled")
    if draw_id:     q = q.eq("draw_id", draw_id)
    if data_inicio: q = q.gte("created_at", data_inicio)
    if data_fim:    q = q.lte("created_at", data_fim + "T23:59:59")
    if user["role"] == "cambista": q = q.eq("user_id", user["id"])
    data = q.execute().data or []

    total_arrecadado = sum(float(a["amount"]) for a in data)
    total_premios    = sum(float(a.get("prize_amount") or 0) for a in data)
    return {
        "total_apostas": len(data),
        "total_arrecadado": total_arrecadado,
        "total_premios": total_premios,
        "saldo": total_arrecadado - total_premios,
        "por_status": {
            "pending": sum(1 for a in data if a["status"] == "pending"),
            "won":     sum(1 for a in data if a["status"] == "won"),
            "lost":    sum(1 for a in data if a["status"] == "lost"),
        }
    }

@router.get("/ranking")
async def relatorio_ranking(
    draw_id: Optional[int] = None,
    data_inicio: Optional[str] = None,
    data_fim: Optional[str] = None,
    user=Depends(get_current_user)
):
    q = supabase.table("apostas").select("amount, prize_amount, user_id") \
        .neq("status", "cancelled")
    if draw_id:     q = q.eq("draw_id", draw_id)
    if data_inicio: q = q.gte("created_at", data_inicio)
    if data_fim:    q = q.lte("created_at", data_fim + "T23:59:59")
    data = q.execute().data or []

    vmap = {}
    for a in data:
        vid = a["user_id"]
        if vid not in vmap:
            vmap[vid] = {"user_id": vid, "total_apostas": 0, "total_arrecadado": 0.0, "total_premios": 0.0}
        vmap[vid]["total_apostas"] += 1
        vmap[vid]["total_arrecadado"] += float(a["amount"])
        vmap[vid]["total_premios"] += float(a.get("prize_amount") or 0)

    ranking = sorted(
        [{"saldo": v["total_arrecadado"] - v["total_premios"], **v} for v in vmap.values()],
        key=lambda x: x["total_arrecadado"], reverse=True
    )
    return ranking

@router.get("/risco")
async def relatorio_risco(draw_id: Optional[int] = None, user=Depends(get_current_user)):
    q1 = supabase.table("risk_exposure").select("*").order("valor_vendido", desc=True)
    q2 = supabase.table("risk_discharge").select("*").order("created_at", desc=True).limit(50)
    if draw_id:
        q1 = q1.eq("draw_id", draw_id)
        q2 = q2.eq("draw_id", draw_id)
    return {"exposure": q1.execute().data or [], "discharge": q2.execute().data or []}

@router.get("/caixa")
async def relatorio_caixa(data: Optional[str] = None, user=Depends(get_current_user)):
    hoje = data or date.today().isoformat()
    inicio = hoje + "T00:00:00"
    fim    = hoje + "T23:59:59"
    res = supabase.table("apostas").select("amount, prize_amount, status") \
        .neq("status", "cancelled").gte("created_at", inicio).lte("created_at", fim).execute()
    rows = res.data or []
    total_bruto   = sum(float(a["amount"]) for a in rows)
    total_premios = sum(float(a.get("prize_amount") or 0) for a in rows)
    return {
        "data": hoje,
        "total_apostas": len(rows),
        "total_bruto": total_bruto,
        "total_premios": total_premios,
        "liquido": total_bruto - total_premios
    }

File: backend/src/routes/relatorios.js
const router = require('express').Router();
const auth = require('../middleware/auth');
const supabase = require('../db');

// GET /relatorios/geral?draw_id=&data=
router.get('/geral', auth, async (req, res) => {
  try {
    const { draw_id, data_inicio, data_fim } = req.query;
    let q = supabase.from('apostas')
      .select('amount, prize_amount, status, bet_type, vendedor_id, draw_id, vendedores(name, commission_rate)')
      .neq('status', 'cancelled');
    if (draw_id) q = q.eq('draw_id', draw_id);
    if (data_inicio) q = q.gte('created_at', data_inicio);
    if (data_fim) q = q.lte('created_at', data_fim + 'T23:59:59');
    if (req.user.role === 'cambista') q = q.eq('vendedor_id', req.user.id);
    const { data, error } = await q;
    if (error) return res.status(400).json({ error: error.message });

    const total_arrecadado = data.reduce((s, a) => s + Number(a.amount), 0);
    const total_premios = data.reduce((s, a) => s + Number(a.prize_amount || 0), 0);
    const total_comissao = data.reduce((s, a) => {
      const rate = a.vendedores?.commission_rate || 0;
      return s + Number(a.amount) * Number(rate) / 100;
    }, 0);
    const saldo = total_arrecadado - total_premios - total_comissao;

    res.json({
      total_apostas: data.length,
      total_arrecadado,
      total_premios,
      total_comissao,
      saldo,
      por_status: {
        pending: data.filter(a => a.status === 'pending').length,
        won: data.filter(a => a.status === 'won').length,
        lost: data.filter(a => a.status === 'lost').length,
      }
    });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// GET /relatorios/ranking?draw_id=
router.get('/ranking', auth, async (req, res) => {
  try {
    const { draw_id, data_inicio, data_fim } = req.query;
    let q = supabase.from('apostas')
      .select('amount, prize_amount, vendedor_id, vendedores(name, commission_rate)')
      .neq('status', 'cancelled');
    if (draw_id) q = q.eq('draw_id', draw_id);
    if (data_inicio) q = q.gte('created_at', data_inicio);
    if (data_fim) q = q.lte('created_at', data_fim + 'T23:59:59');
    const { data, error } = await q;
    if (error) return res.status(400).json({ error: error.message });

    const vendedorMap = {};
    for (const a of data) {
      const vid = a.vendedor_id;
      if (!vendedorMap[vid]) vendedorMap[vid] = {
        vendedor_id: vid, name: a.vendedores?.name || 'N/A',
        total_apostas: 0, total_arrecadado: 0, total_premios: 0, comissao: 0
      };
      vendedorMap[vid].total_apostas++;
      vendedorMap[vid].total_arrecadado += Number(a.amount);
      vendedorMap[vid].total_premios += Number(a.prize_amount || 0);
      vendedorMap[vid].comissao += Number(a.amount) * Number(a.vendedores?.commission_rate || 0) / 100;
    }
    const ranking = Object.values(vendedorMap)
      .map(v => ({ ...v, saldo: v.total_arrecadado - v.total_premios - v.comissao }))
      .sort((a, b) => b.total_arrecadado - a.total_arrecadado);
    res.json(ranking);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// GET /relatorios/risco?draw_id=
router.get('/risco', auth, async (req, res) => {
  try {
    const { draw_id } = req.query;
    let q = supabase.from('risk_exposure').select('*').order('valor_vendido', { ascending: false });
    if (draw_id) q = q.eq('draw_id', draw_id);
    const { data: exposure, error: e1 } = await q;
    let q2 = supabase.from('risk_discharge').select('*').order('created_at', { ascending: false }).limit(50);
    if (draw_id) q2 = q2.eq('draw_id', draw_id);
    const { data: discharge } = await q2;
    if (e1) return res.status(400).json({ error: e1.message });
    res.json({ exposure: exposure || [], discharge: discharge || [] });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

// GET /relatorios/caixa?data=YYYY-MM-DD
router.get('/caixa', auth, async (req, res) => {
  try {
    const { data: dataParam } = req.query;
    const inicio = dataParam ? dataParam + 'T00:00:00' : new Date().toISOString().slice(0,10) + 'T00:00:00';
    const fim = dataParam ? dataParam + 'T23:59:59' : new Date().toISOString().slice(0,10) + 'T23:59:59';
    const { data, error } = await supabase.from('apostas')
      .select('amount, prize_amount, status, vendedor_id, vendedores(name, commission_rate)')
      .neq('status', 'cancelled')
      .gte('created_at', inicio).lte('created_at', fim);
    if (error) return res.status(400).json({ error: error.message });

    const total_bruto = data.reduce((s, a) => s + Number(a.amount), 0);
    const total_premios = data.reduce((s, a) => s + Number(a.prize_amount || 0), 0);
    const total_comissao = data.reduce((s, a) => s + Number(a.amount) * Number(a.vendedores?.commission_rate || 0) / 100, 0);
    res.json({
      data: dataParam || new Date().toISOString().slice(0,10),
      total_apostas: data.length,
      total_bruto,
      total_premios,
      total_comissao,
      liquido: total_bruto - total_premios - total_comissao
    });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

module.exports = router;
